### ISSUE TITLE
Amplify.GraphQLError

### ISSUE THREAD

**Issue by springlo** [aws-amplify/amplify-ios#880]
My request:
```
let object = Case(doctorID: doctorId, status: .open, alphaReport: report, userInfo: User.shared.info, doctor: doctor)
        let build = GraphQLRequest<Case>.create(object)
        let req = GraphQLRequest<Case>(apiName: API.apiName.authApi, document: build.document, variables: build.variables,
                                        responseType: Case.self, decodePath: build.decodePath)
        Amplify.API.mutate(request: req) 
```

got error:
```
GraphQL service returned a successful response containing errors: [Amplify.GraphQLError(message: "The variables input contains a field name \'casedoctor\' that is not defined for input object type \'CreateCaseInput\' ", locations: nil, path: nil, extensions: nil)]
```
When I print build.variables, the key `doctor` not found, and changed to `casedoctor`.

**Other info follows:**

*Model*
```
type Case @model @auth(
  rules: [
    { allow: owner, ownerField: "userID" },
    { allow: owner, ownerField: "doctorID" }
    ]
){
  id: ID!
  userID: ID!
  doctorID: ID!
  status: OrderStatus!
  diagnosis: String
  doctorSigniture: String
  images: [Image] @connection(keyName: "byConnectID", fields: ["id"]) 
  alphaReport: String
  doctorPrescription: String
  userInfo: UserInfo @connection(fields: ["userID"])
  doctor: Doctor @connection(fields: ["doctorID"])
  messages: [Message] @connection(keyName: "byCase", fields: ["id"])
}
```

*Case.swift*
```

public struct Case: Model {
  public let id: String
  public var doctorID: String
  public var status: OrderStatus
  public var diagnosis: String?
  public var doctorSigniture: String?
  public var images: List<Image>?
  public var alphaReport: String?
  public var doctorPrescription: String?
  public var userInfo: UserInfo?
  public var doctor: Doctor?
  public var messages: List<Message>?
  
  public init(id: String = UUID().uuidString,
      doctorID: String,
      status: OrderStatus,
      diagnosis: String? = nil,
      doctorSigniture: String? = nil,
      images: List<Image>? = [],
      alphaReport: String? = nil,
      doctorPrescription: String? = nil,
      userInfo: UserInfo? = nil,
      doctor: Doctor? = nil,
      messages: List<Message>? = []) {
      self.id = id
      self.doctorID = doctorID
      self.status = status
      self.diagnosis = diagnosis
      self.doctorSigniture = doctorSigniture
      self.images = images
      self.alphaReport = alphaReport
      self.doctorPrescription = doctorPrescription
      self.userInfo = userInfo
      self.doctor = doctor
      self.messages = messages
  }
}
```
Case+Schema.swift
```
extension Case {
  // MARK: - CodingKeys 
   public enum CodingKeys: String, ModelKey {
    case id
    case doctorID
    case status
    case diagnosis
    case doctorSigniture
    case images
    case alphaReport
    case doctorPrescription
    case userInfo
    case doctor
    case messages
  }
  
  public static let keys = CodingKeys.self
  //  MARK: - ModelSchema 
  
  public static let schema = defineSchema { model in
    let `case` = Case.keys
    
    model.authRules = [
      rule(allow: .owner, ownerField: "userID", identityClaim: "cognito:username", operations: [.create, .update, .delete, .read]),
      rule(allow: .owner, ownerField: "doctorID", identityClaim: "cognito:username", operations: [.create, .update, .delete, .read])
    ]
    
    model.pluralName = "Cases"
    
    model.fields(
      .id(),
      .field(`case`.doctorID, is: .required, ofType: .string),
      .field(`case`.status, is: .required, ofType: .enum(type: OrderStatus.self)),
      .field(`case`.diagnosis, is: .optional, ofType: .string),
      .field(`case`.doctorSigniture, is: .optional, ofType: .string),
      .hasMany(`case`.images, is: .optional, ofType: Image.self, associatedWith: Image.keys.connectID),
      .field(`case`.alphaReport, is: .optional, ofType: .string),
      .field(`case`.doctorPrescription, is: .optional, ofType: .string),
      .belongsTo(`case`.userInfo, is: .optional, ofType: UserInfo.self, targetName: "userID"),
      .hasOne(`case`.doctor, is: .optional, ofType: Doctor.self, associatedWith: Doctor.keys.id),
      .hasMany(`case`.messages, is: .optional, ofType: Message.self, associatedWith: Message.keys.caseID)
    )
    }
}

```

**Comment by lawmicha** [comment]
Hi @springlo, I took a simplied version of the schema below
```
type Case @model @auth(
  rules: [
    { allow: owner, ownerField: "userID" },
    { allow: owner, ownerField: "doctorID" }
    ]
){
  id: ID!
  userID: ID!
  doctorID: ID!
  userInfo: UserInfo @connection(fields: ["userID"])
  doctor: Doctor @connection(fields: ["doctorID"])
}

type Doctor @model @auth(
    rules: [
        { allow: owner, ownerField: "id" }
    ]
    ){
    id: ID!
    name: String!
}

type UserInfo @model @auth(rules: [
  { allow: owner },
  { allow: private, operations: [read] },
  { allow: public, operations: [read] }
]) {
  id: ID!
  username: String!
}
```

The first issue I encountered with my provisioning is that my `Case+Schema.swift` file does not compile due to the `case` keyword not encapsulated in the backticks. This is an issue with the codegen. I manually updated the `Case+Schema.swift` file, similar to yours.

When you create a Case object, you don't need to pass in the doctor object, only the doctorID is required:

```swift
if let user = Amplify.Auth.getCurrentUser() {
    let case1 = Case(userID: user.username, doctorID: user.username)
    Amplify.API.mutate(request: .create(case1))
}
```

this translates the case object's fields into the graphql variables input and matches what the service expects:
```
// CreateCaseInput
id: ID
userID: ID!
doctorID: ID!
```

If you remove the doctor object from your request, you'll then run into the issue where the nil doctor object is being passed as a GraphQL Input. This is the same bug as described in https://github.com/aws-amplify/amplify-ios/issues/874#issuecomment-731494120

Until this is fixed (by Not adding the doctor key to the graphql input), you'll have to create your own Case mutation request like so:

```swift
extension GraphQLRequest {
    static func createCase(userId: String, doctorId: String) -> GraphQLRequest<Case> {
        let document = """
        mutation CreateCase {
          createCase(input: {doctorID: "\(doctorId)", userID: "\(userId)"}) {
            createdAt
            doctorID
            id
            updatedAt
            userID
          }
        }
        """
        return GraphQLRequest<Case>(document: document, responseType: Case.self, decodePath: "createCase")
    }
}
```

Then you can make the call like
```swift
if let user = Amplify.Auth.getCurrentUser() {
    let username = user.username
    Amplify.API.mutate(request: .createCase(userId: username, doctorId: username))
}
```


## Querying for a Case
You can get a Case by its `id` like so:
```swift
Amplify.API.query(request: .get(Case.self, byId: caseId))
```

One thing to note here is that the generated selection set which determines the level of depth to retrieve is set to 1 by default.  This means the object's doctor and userInfo objects will be nil. We do want to address this in the future by allowing you to specify which connections to retrieve and auto generate the selection set. For now, if you want either of the doctor and the userInfo back from the case, you will have to perform a separate query for the doctor or userInfo or expand the selection set in the GraphQL document like so:

```
extension GraphQLRequest {
    static func queryWithDoctor(caseId: String) -> GraphQLRequest<Case> {
        let document = """
        query GetCaseWithDoctor {
          getCase(id: "\(caseId)") {
            createdAt
            doctorID
            id
            updatedAt
            userID
            doctor {
              createdAt
              id
              name
              updatedAt
            }
          }
        }
        """
        return GraphQLRequest<Case>(document: document, responseType: Case.self, decodePath: "getCase")
    }
}
```

Then you can perform the query by `Amplify.API.query(request: .queryWithDoctor(caseId: caseId))`

I'll keep this issue open to address the two bugs
- Special Swift `case` keyword used for codegen
- Ignoring the hasOne optional Doctor field in create mutations

Hope this helps, please let us know if you are still blocked

**Comment by lawmicha** [comment]
Upon further investigation, my simplified schema does not represent the same issue, however it does bring out the codegen issue related to `hasOne` and `belongsTo`.

When there is a `belongsTo`, the explicit field is removed from the Model. For example, in your Case Model, the belongsTo `userInfo` field contains targetName and `userId` removed from the model resulting in 
```swift
public struct Case: Model {
  public var userInfo: UserInfo?
}

extension Case {
  .belongsTo(`case`.userInfo, is: .optional, ofType: UserInfo.self, targetName: "userID"),
}
```
However, for the `hasOne` relationships, the explicit field is there. Take the Doctor connection for example:
```swift
public struct Case: Model {
  public var doctorID: String
  public var doctor: Doctor?
}

extension Case {
  .hasOne(`case`.doctor, is: .optional, ofType: Doctor.self, associatedWith: Doctor.keys.id),
}
```


I think codegen should remove the explicit field, and add the targetName into the hasOne
```swift
public struct Case: Model {
  public var doctor: Doctor?
}

extension Case {
  .hasOne(`case`.doctor, is: .optional, ofType: Doctor.self, associatedWith: Doctor.keys.id, targetName: "doctorID"),
}
```

With the codegen change, and library changes, then you should be able to successfully create a mutation as that doctor object will get translated to the GraphQLRequest by using the targetName as 
```swift
{
 "doctorID": doctor.id
}
```

**Comment by lawmicha** [comment]
Hi @springlo thanks for your patience with this, the code changes have been released in [1.5.1](https://github.com/aws-amplify/amplify-ios/releases/tag/v1.5.1). To get these changes, 
1. please update your pods - `pod install --repo-update` 
2. Update Amplify CLI `amplify upgrade`. This upgraded mine to `4.40.0`
3. re-run `amplify codegen models`. You should notice that your model schemas with `hasOne` relationships will contain the targetName, ie.
```
.hasOne(`case`.userInfo, is: .optional, ofType: UserInfo.self, associatedWith: UserInfo.keys.id, targetName: "userID"),
.hasOne(`case`.doctor, is: .optional, ofType: Doctor.self, associatedWith: Doctor.keys.id, targetName: "doctorID")
```
I also had to encase the "case" keyword with backticks in the swift file for it to compile successfully and do a quick test. Let us know if this works for you.

**Comment by lawmicha** [comment]
Will be closing this since the "hasOne with targetName" issue has been released in both the library and CLI. Feel free to let us know if you have any questions regarding the fix or please open a new issue if you have other problems
